Core/ALife.Core/WorldObjects/Agents/Brains/NeuralNetworkBrain.cs

```diff
@@ -28,217 +31,318 @@ public NeuralNetworkBrain(Agent self, List<int> layers)
                 }
             }
-            if(actionNeurons.Count != templateLayer.Neurons.Count)
+            doubleInfo.AddRange(dendriteWeights);
+
+            List<string> NeuronNames = new List<string>();
+            for(int m = 0; m < layerCount; ++m)
             {
-                throw new NotImplementedException("Cannot clone a brain with a different number of actions");
+                NeuronNames.AddRange(Layers[m].Neurons.Select(neu => neu.Name));
             }
-            Layer actionLayer = new Layer(actionNeurons.Count);
-            actionLayer.Neurons.AddRange(actionNeurons);
-            return actionLayer;
+
+            StringBuilder sb = new StringBuilder();
+            //sb.Append($"[{string.Join(",", layerInfoArray)}]");
+            //sb.AppendLine($"[{string.Join(",", doubleInfo.ToArray())}]");
+            sb.AppendLine(ConvertBrainArraysToStr(layerInfoArray, doubleInfo.ToArray()));
+            sb.AppendLine(String.Join(",", NeuronNames));
+
+            return sb.ToString();
         }
 
-        private static Layer CreateSenseLayer(Agent self)
+        private string ConvertBrainArraysToStr(byte[] layerInfo, double[] allOtherInfo)
         {
-            List<FuncNeuron> senseNeurons = new List<FuncNeuron>();
-            foreach(SenseCluster sc in self.Senses)
+            int shortLength = layerInfo.Length * sizeof(byte);
+            int doubleLength = allOtherInfo.Length * sizeof(double);
+
+            byte[] asBytes = new byte[shortLength + doubleLength];
+            Buffer.BlockCopy(layerInfo, 0, asBytes, 0, shortLength);
+            Buffer.BlockCopy(allOtherInfo, 0, asBytes, shortLength, doubleLength);
+            string s = Convert.ToBase64String(asBytes);
+
+            return s;
+        }
+
+        private NeuralNetworkBrainImport ExtractBrainInfoFromStr(string exportedBrain)
+        {
+            string[] lines = exportedBrain.Split(Environment.NewLine.ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+            string numeric = lines[0];
+
+            NeuralNetworkBrainImport output = new NeuralNetworkBrainImport();
+
+            byte[] inputBytes = Convert.FromBase64String(numeric);
+            
+            byte layerCount = inputBytes[0];
+            output.LayerCount = layerCount;
+            output.NeuronCounts = new ArraySegment<byte>(inputBytes, 1, layerCount).ToArray();
+
+            output.NeuronBiases = new double[layerCount][];
+            output.DendriteWeights = new double[layerCount][][];
+
+            int streamCursor = 1 + layerCount; //This is in bytes because the preceding information are bytes!!! 
+
+            double[] modStats = new double[2];
+            Buffer.BlockCopy(inputBytes, streamCursor, modStats, 0, 2 * sizeof(double));
+            streamCursor += 2 * sizeof(double);
+            output.ModStats = modStats;
+
+            for(int i = 0; i < layerCount; ++i)
```

Comment:
Like above, I'd probably either have in a single loop for everything if we can, or move these into helper methods (preferred).

Reply:
Single loop would break the output array order.